cpp/src/parquet.cpp

```diff
@@ -124,11 +146,77 @@ struct create_result_store_fn {
+std::vector<int> GetArrowColumnIndices(const std::string& file_path,
+                                       const std::vector<std::string>& columns)
+{
+  // Open the file and get the schema
+  auto input = ARROW_RESULT(arrow::io::ReadableFile::Open(file_path));
+  std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
+  auto status = parquet::arrow::OpenFile(input, arrow::default_memory_pool(), &arrow_reader);
+  std::shared_ptr<arrow::Schema> schema;
+  status = arrow_reader->GetSchema(&schema);
+
+  // Find the column indices
+  std::vector<int> column_indices;
+  column_indices.reserve(columns.size());
+  for (const auto& col : columns) {
+    auto idx = schema->GetFieldIndex(col);
+    if (idx < 0) { throw std::runtime_error("Column " + col + " not found in Parquet schema."); }
+    column_indices.push_back(idx);
+  }
+  return column_indices;
+}
+
 class ParquetRead : public Task<ParquetRead, OpCode::ParquetRead> {
  public:
   static inline const auto TASK_CONFIG =
     legate::TaskConfig{legate::LocalTaskID{OpCode::ParquetRead}};
 
+  static void cpu_variant(legate::TaskContext context)
+  {
+    TaskContext ctx{context};
+    const auto file_paths        = argument::get_next_scalar_vector<std::string>(ctx);
+    const auto columns           = argument::get_next_scalar_vector<std::string>(ctx);
+    const auto ngroups_per_file  = argument::get_next_scalar_vector<size_t>(ctx);
+    const auto nrow_groups_total = argument::get_next_scalar<size_t>(ctx);
+    PhysicalTable tbl_arg        = argument::get_next_output<PhysicalTable>(ctx);
+    argument::get_parallel_launch_task(ctx);
+
+    if (file_paths.size() != ngroups_per_file.size()) {
+      throw std::runtime_error("internal error: file path and nrows size mismatch");
+    }
+
+    auto [my_groups_offset, my_num_groups] =
+      evenly_partition_work(nrow_groups_total, ctx.rank, ctx.nranks);
+
+    if (my_num_groups == 0) {
+      tbl_arg.bind_empty_data();
+      return;
+    }
+
+    auto column_indices = GetArrowColumnIndices(file_paths.at(0), columns);
+    // Iterate over files
+    auto [files, row_groups] =
+      find_files_and_row_groups(file_paths, ngroups_per_file, my_groups_offset, my_num_groups);
+    std::vector<std::shared_ptr<arrow::Table>> tables;
+    for (int i = 0; i < files.size(); i++) {
+      auto input = ARROW_RESULT(arrow::io::ReadableFile::Open(files[i]));
+      std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
+      auto status = parquet::arrow::OpenFile(input, arrow::default_memory_pool(), &arrow_reader);
+      std::unique_ptr<arrow::RecordBatchReader> batch_reader;
+      status = arrow_reader->GetRecordBatchReader(row_groups[i], column_indices, &batch_reader);
+      tables.push_back(ARROW_RESULT(batch_reader->ToTable()));
+    }
+    // Concatenate the tables
+    if (tables.size() == 0) {
+      tbl_arg.bind_empty_data();
+    } else {
+      tbl_arg.move_into(std::move(ARROW_RESULT(arrow::ConcatenateTables(tables))));
```

Comment:
I do wonder if it's worth adding a `move_into()` (if we go ahead with something like the other PR) that does the concatenation part and copies exactly once.

That way, we could use an eager allocation, and just always copy.  And we could even read one row group at a time and then copy (or similar).

Anyway, nothing to do here!